src/bucket/BucketApplicator.cpp

```diff
@@ -134,7 +134,22 @@ BucketApplicator::advance(BucketApplicator::Counters& counters)
             }
             else
             {
-                ltx->eraseWithoutLoading(e.deadEntry());
+                if (protocolVersionIsBefore(
```

Comment:
I think this doesn't handle every case, but I'm not 100% sure. Here's the scenario I'm imagining:

1. Bucket on level 9 contains `DEADENTRY` with `ledgerVersion = 0` (eg. it's an old bucket)
2. Bucket on level 2 contains `LIVEENTRY` with `ledgerVersion = 0` (eg. it's an old bucket)
3. Bucket on level 1 contains `DEADENTRY` with `ledgerVersion = 11` (eg. it's a new bucket)

Then the bucket on level 1 merges into the bucket on level 2, yielding a bucket containing `DEADENTRY` with `ledgerVersion = 11` as implemented in https://github.com/stellar/stellar-core/blob/ea540894ac5e4bd72c30bbf392099b6a279c50b0/src/bucket/Bucket.cpp#L358-L366

If we then applied this bucket list, we would still get `DEADENTRY` merging into `DEADENTRY`.

It's definitely possible I have a misunderstanding here. Let me know what you think. I'm concerned there might be a related issue on lines 106,122.

Reply:
So you're saying that if you have a delete and a subsequent create in separate pre V11 buckets, a delete in a V11 bucket can cause this issue. I think you're right, but I'll need to look at the bucket code to get a better understanding. 

If we change the check from V11 to V12, would all of these dead and live entries in different buckets have been merged into level 10 with a final live entry state? If I understand the comments in `BucketList.h', the top level bucket is merged every two million ledgers, which is less than the gap between V11 and V12.

Reply:
I agree @jonjove is right here. Another way of thinking about this bug class is to note that bucket apply is making judgments about the protocol version of the bucket it is applying, but not the older buckets already applied. And the new v11 DEADENTRY semantics (that 2 adjacent DEADENTRYs is an error) only holds in the _newer_ direction from the v10/v11 boundary.

Reply:
Yes so basically we have to default to the "don't trust older buckets" path if any of the buckets is v10 or older; so this issue needs to also be fixed for the update/create case because older buckets appear (from v11's point of view) to have wrong or missing data because they got constructed differently.

Reply:
PR updated

Reply:
I added a commit that applies live entries as init for buckets at the oldest level.